mesonbuild/interpreter.py

```diff
@@ -2784,13 +2784,50 @@ def run(self):
         super().run()
         mlog.log('Build targets in project:', mlog.bold(str(len(self.build.targets))))
 
+
+    # Check that the indicated file is within the same subproject
+    # as we currently are. This is to stop people doing
+    # nasty things like:
+    #
+    # f = files('../../master_src/file.c')
+    #
+    # Note that this is validated only when the file
+    # object is generated. The result can be used in a different
+    # subproject than it is defined in (due to e.g. a
+    # declare_dependency).
+    def validate_within_subproject(self, subdir, fname):
+        norm = os.path.normpath(os.path.join(subdir, fname))
+        if os.path.isabs(norm):
+            if not norm.startswith(self.environment.source_dir):
+                # Grabbing files outside the source tree is ok.
+                # This is for vendor stuff like:
+                #
+                # /opt/vendorsdk/src/file_with_license_restrictions.c
+                return
+            rel_norm = norm[len(self.environment.source_dir):]
+            assert(not os.path.isabs(rel_norm))
+            self.validate_within_subproject(*os.path.split(rel_norm))
+            return
```

Comment:
It seems to me that instead of calling the function again here and returning, you could do:

```
    norm = norm[len(self.environment.source_dir):]
    assert(not os.path.isabs(norm))
segments = norm.split(os.path.sep)
...
```

This effectively means the 'isabs' if condition converts an absolute path to the source tree into a relative path, which is clearer and equivalent, I think.

Reply:
updated